File: profiling/RandomReadsList.java

import java.util.concurrent.ThreadLocalRandom;
import org.python.types.Int;
import org.python.types.List;

public class RandomReadsList {
    private static final int LIST_SIZE = 1000000;
    private static final int NR_READS = 1000000;
    private static final long MAX_VALUE = Long.MAX_VALUE;
    private static final long MIN_VALUE = Long.MIN_VALUE;

    public static void main(String[] args) {
        List l = new List();
        for (int i = 0; i < LIST_SIZE; ++i) {
            long rand = ThreadLocalRandom.current().nextLong(MIN_VALUE, MAX_VALUE);
            l.append(Int.getInt(rand));
        }

        long start = System.nanoTime();

        for (int i = 0; i < NR_READS; ++i) {
            ThreadLocalRandom.current().nextLong(MIN_VALUE, MAX_VALUE);
            org.python.Object tmp = l.__getitem__(Int.getInt((int) (Math.random() * LIST_SIZE)));
        }

        System.out.println("Time taken: " + (System.nanoTime() - start) / 1000000000.);
    }

}

File: profiling/random_reads_list.py
import random
import sys
import time

l = list()
LIST_SIZE = 100000
MININT = -sys.maxsize - 1

for i in range(LIST_SIZE):
    l.append(random.randint(MININT, sys.maxsize))

AMOUNT_OF_READS = 1000000

start_time = time.time()

for i in range(AMOUNT_OF_READS):
    tmp = l[random.randint(0, LIST_SIZE - 1)]

print("Time taken: ", time.time() - start_time)
